src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -536,6 +536,31 @@ public async Task OnLocationChangedAsync(string uri, string state, bool intercep
         }
     }
 
+    public async Task OnLocationChangingAsync(int callId, string uri, bool intercepted)
+    {
+        AssertInitialized();
+        AssertNotDisposed();
+
+        try
+        {
+            var shouldContinueNavigation = await Renderer.Dispatcher.InvokeAsync(async () =>
+            {
+                Log.LocationChanging(_logger, uri, CircuitId);
+                var navigationManager = (RemoteNavigationManager)Services.GetRequiredService<NavigationManager>();
+                return await navigationManager.HandleLocationChangingAsync(uri, intercepted);
+            });
+
+            await Client.SendAsync("JS.EndLocationChanging", callId, shouldContinueNavigation);
+        }
+        catch (Exception ex)
+        {
+            // Exceptions thrown by location changing handlers should be treated like unhandled exceptions in user-code.
```

Comment:
I wonder if we should really be treating exceptions thrown in location changing handlers as non-fatal? In the current WebAssembly implementation of this feature, the navigation is canceled if an exception is thrown in the location changing handler. Should we treat it as a critical failure there instead? And, if we decide to make this case non-fatal, should we let the navigation continue if no other handlers cancel it?

Reply:
I'd err on the side of treating unhandled exceptions as fatal unless there's a definite use case in which developers can't be expected to avoid unhandled exceptions. In this case I expect a developer could use try/catch inside their handler if they are doing something that might fail.

Generally in WebAssembly we've had a more laissez-faire attitude because we're inheriting the JS semantics around errors being something you log but may be able to continue afterwards. On Server, it's much more important to have clearly defined rules about when you can and can't continue after an exception.

> Should we treat it as a critical failure there instead?

If possible, that would be nice. But as long as something ends up in the JS console giving the exception details, that's enough. It just has to be reasonable for the developer to see what went wrong.